### PR TITLE
Login script

### PR DESCRIPTION
**Pull request by mbogor**
#### What's this PR do?
refactors the login script
adds a refresh cache cronjob
turns cache into a hash table from an array to improve search efficiency
#### How should this be manually tested?
node runJobs and node www/app.js
#### What are the relevant tickets?
https://admin.crisistextline.org/jira/browse/INT-421

### CODE REVIEW

File: cache.js [hunk 1]
```diff
@@ -9,7 +9,11 @@ function WiWUsersCache(){
       if(response.message) reject('Call to get all WiW users failed');
       try {
         if(typeof response !=='object') throw('WiW all users response is not in the proper JSON format');
-        resolve(response.users);
+        var hash = {};
+        response.users.forEach(function(user){
+          hash[user.email] = user;
```
Comment (apackin) [inline cache.js]: 👍

File: cache.js [hunk 2]
```diff
@@ -9,7 +9,11 @@ function WiWUsersCache(){
       if(response.message) reject('Call to get all WiW users failed');
       try {
         if(typeof response !=='object') throw('WiW all users response is not in the proper JSON format');
-        resolve(response.users);
+        var hash = {};
```
Comment (apackin) [inline cache.js]: const

File: jobs/cacheRefresh.js [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+new CronJob(CONFIG.time_interval.cache_refresh, refresh, null, true);
+
+function refresh(){
+  cache()
+  .then(function(users){
+    global.USERS_CACHE = users;
+    var test = Object.keys(global.USERS_CACHE)[0];
+    CONSOLE_WITH_TIME(test);
+    CONSOLE_WITH_TIME(users[test]);
```
Comment (apackin) [inline jobs/cacheRefresh.js]: What is the purpose of these CONSOLE_WITH_TIME's ?

File: jobs/cacheRefresh.js [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+    CONSOLE_WITH_TIME('The cache has been refreshed')
+  })
+  .catch(function(error){
+    CONSOLE_WITH_TIME('error showing up from cache initialization');
+    console.error(error);
```
Comment (apackin) [inline jobs/cacheRefresh.js]: It's a great idea to start using console.error, should we make a global ERROR_WITH_TIME function and use this everywhere we are logging errors?

File: jobs/cacheRefresh.js [hunk 5]
```diff
@@ -0,0 +1,22 @@
+'use strict';
```
Comment (apackin) [inline jobs/cacheRefresh.js]: This function is very similar to CacheInit in runJobs. Could we merge them somehow?

File: jobs/scheduling/removePending.js [hunk 6]
```diff
@@ -33,17 +33,15 @@ function getPendingUsers(){
 function getDeletedUsers(pending){
 	return new Promise(function(resolve, reject){
 		let reactivate = [];
-		global.USERS_CACHE.forEach(function(user){
-			if (!user.is_deleted && pending[user.email]) {
-				// reject the pending user
-				rejectUser(pending[user.email]);
+		for(var key in global.USERS_CACHE){
```
Comment (apackin) [inline jobs/scheduling/removePending.js]: I think the logic in these if's can be combines. Something like:
if(pending[key]) {
    if(global.USERS_CACHE[key].is_deleted) reactivate.push(pending[key]);   
    rejectUser(pending[key]);
}

File: runJobs.js [hunk 7]
```diff
@@ -10,14 +11,17 @@ if (process.env.NODE_ENV !== 'production') {
...
-	cache()
-	.then(function(users){
-		global.USERS_CACHE = users;
-		const jobs = require('require-all')(__dirname + '/jobs');
-	})
-	.catch(function(error){
-		console.error(error);
-		CacheInit();
-	})
+  if (global.USERS_CACHE) return;
```
Comment (apackin) [inline runJobs.js]: This functionality exists in jobs/cacheRefresh.js

File: test/recurOpenShifts.js [hunk 8]
```diff
@@ -31,7 +31,7 @@ describe('recurOpenShifts', function() {
   describe('returnMaxOpenShiftCountForTime', function() {
 
     it('should return max open shift count for a time', function () {
-      assert.equal(recurOpenShifts.returnMaxOpenShiftCountForTime(targetTime), 19);
+      assert.equal(recurOpenShifts.returnMaxOpenShiftCountForTime(targetTime), 29);
```
Comment (apackin) [inline test/recurOpenShifts.js]: I thought we already did this? But still :-)

File: www/scheduling/index.js [hunk 9]
```diff
@@ -97,38 +97,32 @@ function reactivate(user){
 
 // retrieves all WiW users and checks for a particular user
 function allUsers(email, altEmail, callback){
-  var match = false;
   var users = global.USERS_CACHE;
   CONSOLE_WITH_TIME('users cache length within allUsers: ', users.length);
```
Comment (apackin) [inline www/scheduling/index.js]: users is an object. @ .length

**Comment by apackin** [comment]
Very smart PR. More efficient and easier to understand 👍  💯

File: jobs/cacheRefresh.js (hunk 5 above)
Comment (mbogor) [inline jobs/cacheRefresh.js]: yas

File: jobs/scheduling/removePending.js (hunk 6 above)
Comment (mbogor) [inline jobs/scheduling/removePending.js]: oh true, did it, thanks

File: jobs/scheduling/removePending.js [hunk 10]
```diff
@@ -22,9 +22,9 @@ function go(){
 function getPendingUsers(){
 	return new Promise(function(resolve, reject){
 		const pending = {};
-		global.USERS_CACHE.forEach(function(user){
-			if(!user.is_active) pending[user.email] = user;
-		});
+		for(let key in global.USERS_CACHE){
+			if(!global.USERS_CACHE[key].is_active) pending[key] = global.USERS_CACHE[key];
```
Comment (tongxiang) [inline jobs/scheduling/removePending.js]: unnecessary to specify `global.USERS_CACHE`

can just write `USERS_CACHE`

**Comment by tongxiang** [comment]
👍

File: jobs/scheduling/removePending.js (hunk 10 above)
Comment (mbogor) [inline jobs/scheduling/removePending.js]: u right

**Pull request merged**